### ISSUE TITLE
How to do floor and ceiling lookups?

### ISSUE THREAD

**Issue by zardus** [grantjenks/sorted_containers#87]
With bintrees being discontinued, and projects using it being urged to migrate to sortedcontainers, I have a feature question. In bintrees, you can look up the "floor" and "ceiling" items given a key that is either too big or too small. For example:

```
import bintrees
x = bintrees.BinaryTree()
x[10] = 'ten'
assert x.floor_item(15) == (10, 'ten')
assert x.ceiling_item(5) == (10, 'ten')
```

As far as I can tell, this isn't possible in sortedcontainers. Am I missing anything? If not, is it a planned feature to add? Alternatively, do you have any pointers for a direction toward implementing it?

**Comment by grantjenks** [comment]
Do you need the prev/succ functionality of bintrees as well?

**Comment by zardus** [comment]
That'd definitely be very useful! Let me know if I can help.

**Comment by grantjenks** [comment]
prev/succ is pretty similar to floor/ceil. I'll try to get to it this week. I'll probably spell them:

* SortedList.floor(value)
* SortedList.ceil(value)
* SortedListWithKey.floor_key(key)
* SortedListWithKey.ceil_key(key)

Actually, you can do the same with the existing "irange" method.

I've got to run now but I'll try to find some time this week to think about it.

**Comment by grantjenks** [comment]
Actually, I now remember I added irange for these use cases. The performance will be excellent if you’re calling prev/succ repeatedly. It also covers the floor/ceil cases. Can you look at that method and see if will work? It’s on sorted list and sorted dict. The performance may not be quite so fast as above but better to optimize after profiling.

**Comment by zardus** [comment]
Awesome, that seems to work! This is the code:

```
import sortedcontainers
s = sortedcontainers.SortedDict()
s[10] = 'ten'
s[3] = 'three'

# floor
assert next(s.irange(maximum=15, reverse=True)) == 10

# ceiling
assert next(s.irange(minimum=5, reverse=False)) == 10

# succ
assert next(s.irange(minimum=3, inclusive=(False, False), reverse=False)) == 10

# prev
assert next(s.irange(maximum=10, inclusive=(False, False), reverse=True)) == 3
```

Awesome stuff! It might be worth making some aliases for this functionality (and, like you say, specializing them if performance is an issue)?

**Comment by grantjenks** [comment]
That looks right. Glad it works!

I benchmarked the "floor_key" operation with bintrees.BinaryTree, sortedcontainers.SortedDict and CustomSortedDict. Here's the timing results:

| size | bintrees.BinaryTree | sortedcontainers.SortedDict | CustomSortedDict |
| ---- | ---- | ---- | ---- |
| 10 | 6.771e-07 | 1.245e-06 | 4.63e-07 |
| 100 | 1.372e-06 | 1.629e-06 | 7.93e-07 |
| 1000 | 1.98e-06 | 3.528e-06 | 1.171e-06 |
| 10000 | 2.06e-06 | 1.093e-05 | 8.759e-07 |
| 100000 | 2.96e-06 | 9.562e-06 | 9.089e-07 |
| 1000000 | 3.01e-06 | 2.979e-05 | 1.004e-06 |

The benchmark constructs a sorted dictionary with "size" random float keys and then looks one up at random. SortedDict is currently between 1-10x slower than bintrees.BinaryTree. The CustomSortedDict is 1-3x faster.

Here's the code:

```python
from __future__ import print_function

import random
import timeit
import bintrees
import sortedcontainers

from test import CustomSortedDict


def median(values):
    return sorted(values)[len(values) // 2]


def run(statement, variable):
    statement = statement.format(variable=variable)
    setup = 'from __main__ import {variable}'.format(variable=variable)
    times = timeit.repeat(statement, setup, repeat=5, number=1000)
    return median(times) / 1000


tree = None
sd = None
csd = None


def benchmark(size):
    global tree, sd, csd

    random.seed(0)
    tree = bintrees.BinaryTree()
    sd = sortedcontainers.SortedDict()
    csd = CustomSortedDict()

    keys = [random.random() for num in range(size)]

    for num, key in enumerate(keys):
        tree[key] = num
        sd[key] = num
        csd[key] = num

    key = keys[0] + 0.001
    tree_time = run('{variable}.floor_key(%r)' % key, 'tree')
    statement = 'next({variable}.irange(maximum=%r, reverse=True))' % key
    sd_time = run(statement, 'sd')
    csd_time = run('{variable}.floor_key(%r)' % key, 'csd')

    print(
        size,
        format(tree_time, '.4g'),
        format(sd_time, '.4g'),
        format(csd_time, '.4g'),
    )


if __name__ == '__main__':
    print(
        'size',
        'bintrees.BinaryTree',
        'sortedcontainers.SortedDict',
        'CustomSortedDict',
    )
    for exp in range(1, 7):
        benchmark(10 ** exp)
```

**Comment by zardus** [comment]
Interesting... It's weird that the runtime doesn't seem to get progressively slower with a bigger list size. Are you running on pypy?

Would you recommend that we adapt something like the CustomSortedDict, or do you think that functionality should make it into sortedcontainer?

**Comment by grantjenks** [comment]
It's not that strange if you look at how "SortedList._islice" is implemented. The iterator returned is optimized for iterating. It's not optimized for returning the first element as quickly as possible. That can/should be changed.

I would adapt to irange and expect that in the next week/month, it will become faster for your use-case :)

Only after profiling would I use the specialized CustomSortedDict.

**Comment by grantjenks** [comment]
New changes at 6321f22. Here's the benchmark results:

 | size | bintrees.BinaryTree | sortedcontainers.SortedDict | CustomSortedDict | 
 | ---- | ---- | ---- | ---- | 
 | 10 | 5.678e-07 | 1.714e-06 | 5.401e-07 | 
 | 100 | 1.206e-06 | 1.85e-06 | 8.721e-07 | 
 | 1000 | 1.613e-06 | 1.998e-06 | 9.54e-07 | 
 | 10000 | 1.714e-06 | 3.684e-06 | 9.983e-07 | 
 | 100000 | 2.316e-06 | 3.919e-06 | 1.001e-06 | 
 | 1000000 | 2.439e-06 | 4.214e-06 | 1.164e-06 | 

The irange method is now 1-2x slower than the floor/ceiling functions but maintains its fast iteration. It also has more consistent performance because it "lazily" slices the lists in C-code.

I think this is "fast-enough" for now. If you need more performance then I'd suggest the CustomSortedDict changes.

I will plan to deploy these improvements in the next week or so.

**Comment by grantjenks** [comment]
Released at v1.5.10 to pypi.org
